cincinnati/src/plugins/internal/arch_filter.rs

```diff
@@ -0,0 +1,356 @@
+        regex::Regex::new(&ARCH_VALIDATION_REGEX_STR).expect("could not create regex");
+}
+
+impl InternalPlugin for ArchFilterPlugin {
+    fn run_internal(self: &Self, internal_io: InternalIO) -> AsyncIO<InternalIO> {
+        let arch = {
+            let interesting_params =
+                try_get_multiple_values!(internal_io.parameters, "arch", "version");
+
+            infer_version(
+                interesting_params.0.map(std::string::ToString::to_string),
+                interesting_params.1.map(std::string::ToString::to_string),
+                self.default_arch.clone(),
+                self.default_arch_threshold_version.clone(),
+            )
+        }
+        .map_err(failure::Error::from);
+
+        let future_result = futures::future::result(arch)
+            .join(futures::future::ok::<_, failure::Error>((
+                internal_io,
+                self.key_prefix.to_owned(),
+                self.key_suffix.to_owned(),
+            )))
+            .and_then(|(arch, (internal_io, key_prefix, key_suffix))| {
+                let mut graph = internal_io.graph;
+
+                // iterate over all releases which have the arch metadata set and
+                // 1. for every release which matches the given `arch`, rewrite the
+                //    arch metadata to only contain `arch`
+                // 2. collect all other releases to be removed
+                let to_remove = {
+                    graph
+                        .find_by_fn_mut(|release| {
+                            match release {
+                                crate::Release::Concrete(concrete_release) => concrete_release
+                                    .metadata
+                                    .get_mut(&format!("{}.{}", key_prefix, key_suffix))
+                                    .map_or(true, |values| {
+                                        if values.split(',').any(|value| value.trim() == arch) {
+                                            *values = arch.clone();
+                                            false
+                                        } else {
+                                            true
+                                        }
+                                    }),
+                                // remove if it's not a ConcreteRelease
+                                _ => true,
+                            }
+                        })
+                        .into_iter()
+                        .map(|(release_id, version)| {
+                            trace!("queuing '{}' for removal", version);
+                            release_id
+                        })
+                        .collect()
+                };
+
+                // remove all matches from the Graph
+                let removed = graph.remove_releases(to_remove);
+
+                trace!("removed {} releases", removed);
+
+                // remove the build suffix from the version
```

Comment:
I'd rather never have arch suffixes on the version strings.  That information can be extracted from the image and placed in release metadata and never be stuffed onto the signed version string.  Release _tags_ (which are outside the image and not signed) can include the arch if release admins want, but that shouldn't matter to Cincinnati.

Reply:
> I'd rather never have arch suffixes on the version strings.

The way the cincinnati library is written requires all version strings to be unique, hence I chose to encode the arch in it.



> That information can be extracted from the image and placed in release metadata and never be stuffed onto the signed version string.

That's what the graph-builder/registry code is doing.

>  Release _tags_ (which are outside the image and not signed) can include the arch if release admins want, but that shouldn't matter to Cincinnati.

It doesn't matter the way the graph-builder is implemented. As you mentioned earlier and I'm repeating now for completeness, we extract the architecture from the manifest, the tag is an opaque string to Cincinnati.